Ew2d qupdate

ludwig-ahrens opened the pull request:
Fix charge_update is now responsible for managing the elastance matrix and the b vector.
Reading and writing of the matrix and vector is done in the fix.
For the writing, the entries are sorted by groups in all other cases the entries are sorted by tags only. This allows an easier communication with the computes.
The computes use only one group, comprising all electrode atoms, the fix gets a group for each electrode with the target potential.
Writing the uninverted capacitance matrix is no longer possible.
I hope you do not mind std::vector. It makes a lot of things shorter and safer.
A test is available in the 'charge-update' branch of the crash-test repo. 
Still not tested for a moving electrolyte.

ludwig-ahrens committed:
Charge udpate working for one timestep

fix_charge_update adapts charges of its group. Without moving atoms the
charges are set to the correct value after the first timestep.
Currently only works with the computes in USER_CONP, which have been
adapted to take one group as argument, only. The potentials are set for
the fix.
Example for a group ele that comprises the groups bot and top:
compute mat ele conp/matrix 0 1.805 file A.mat file_inv capacitance.mat
compute vec ele conp/vector 0 1.805 file vec.cs
fix fxupdate ele charge_update bot -0.5 top 0.5 c_vec c_mat

ludwig-ahrens committed:
Recalculate cs and sn when necessary

For compute_vector* and compute_matrix* in ewald_conp, cs and sn have
to be reacalculated after the atoms have moved.
Fix parallelization bugs of compute_conp_vector

ludwig-ahrens committed:
Write to file from fix charge_update

Write elastance matrix and b vector from fix charge_update instead of
computes.
Writing the capacitance matrix is no longer possible.

ludwig-ahrens committed:
Option to read elastance matrix in charge_update

ludwig-ahrens committed:
Merge remote-tracking branch 'origin/ew2d' into ew2d-qupdate

robeme commented on src/USER-CONP/compute_conp_matrix.cpp:
```diff
@@ -92,22 +89,13 @@ ComputeConpMatrix::ComputeConpMatrix(LAMMPS *lmp, int narg, char **arg)
   // TODO recalculate coulomb matrix every recalc_every
 
-  recalc_every = utils::inumeric(FLERR, arg[4], false, lmp);
+  recalc_every = utils::inumeric(FLERR, arg[3], false, lmp);
```
Das Feature ist obsolet. Im Prinzip wird ja immer wenn der Compute angefordert wird, die Matrix berechnet. recalc_every kann also komplett entfernt werden.

robeme commented on src/USER-CONP/compute_conp_matrix.cpp:
```diff
@@ -266,44 +245,19 @@ void ComputeConpMatrix::init() {
   igroupnum = group->count(igroup);
-  jgroupnum = group->count(jgroup);
-  ngroup = igroupnum + jgroupnum;
+  ngroup = igroupnum;
```
Warum machst du das so? Du könntest doch direkt igroupnum in ngroup umbennen. igroupnum brauchen wir ja eigentlich nicht mehr oder?

robeme commented on src/USER-CONP/compute_conp_matrix.cpp:
```diff
@@ -563,55 +519,32 @@ void ComputeConpMatrix::matrix_assignment() {
-  gradQ_V_inv = new double[ngroup * ngroup];
   gradQ_V = new double *[ngroup];
-  for (bigint i = 0; i < ngroup; i++) gradQ_V[i] = new double[ngroup];
+  array = new double *[ngroup];
```
Wozu benutzt du genau array? Der Name ist etwas nichtsagend oder muss das so sein?

ludwig-ahrens replied:
Das ist muss so, weil es in public von compute.h ist. D.h. es einfach vom fix darauf zuzugreifen.

robeme commented on src/USER-CONP/compute_conp_vector.cpp:
```diff
@@ -55,28 +55,18 @@ ComputeConpVector::ComputeConpVector(LAMMPS *lmp, int narg, char **arg)
   // TODO recalculate coulomb vector every recalc_every
 
-  recalc_every = utils::inumeric(FLERR, arg[4], false, lmp);
+  recalc_every = utils::inumeric(FLERR, arg[3], false, lmp);
```
recalc_every kann auch hier komplett raus. Siehe mein Kommentar dazu in compute_matrix.

robeme commented on src/USER-CONP/compute_conp_vector.cpp:
```diff
@@ -217,37 +194,25 @@ void ComputeConpVector::init() {
-
-  matrix_assignment();
-
+  create_taglist();
```
Vielleicht lieber vector_assignment nennen?

robeme commented on src/USER-CONP/compute_conp_vector.cpp:
```diff
@@ -217,37 +194,25 @@ void ComputeConpVector::init() {
 /* ---------------------------------------------------------------------- */
 
 void ComputeConpVector::compute_vector() {
+  update_mpos();
```
Mit create_taglist erzeugst du die Liste um den Vektor zu adressieren und mit update_mpos pflegst du sie?

ludwig-ahrens replied:
taglist bleibt unverändert. update_mpos() passt mpos an die Atome an, die dem Prozessor zugeordnet sind.

robeme commented on src/USER-CONP/compute_conp_vector.cpp:
```diff
@@ -305,94 +270,49 @@ void ComputeConpVector::pair_contribution() {
+/* ---------------------------------------------------------------------- */
 
+void ComputeConpVector::update_mpos() {
+  int const nall = atom->nlocal + atom->nghost;
```
Für nall kannst du nmax nehmen. Das ist in atom->nmax und umfasst genau die Atome aus nlocal+nghost. Ist etwas sauberer, wenn du das so machst.

robeme commented on src/USER-CONP/ewald_conp.h:
```diff
@@ -85,6 +85,9 @@ class EwaldConp : public KSpace {
   void deallocate_groups();
+ 
+ private:
+  int eikr_step;
```
Was macht das genau?

ludwig-ahrens replied:
compute_vector() und so müssen eik_dot_r() aufrufen, aber nur wenn das in dem Zeitschritt noch nicht gemacht wurde, weil es z.B.  compute_vector_corr() aufgerufen haben könnte.

robeme replied:
eikr_step speichert also den Zeitschritt an dem das letzte Mal eik_dot_r gemacht wurde?

ludwig-ahrens replied:
Ja, die compute()-Methode benutzt das aber nicht, weil die das ja ein zweites Mal ausführen muss.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -0,0 +1,402 @@
+/* ---------------------------------------------------------------------- */
+
+std::vector<int> FixChargeUpdate::local_to_matrix() {
+  int const nall = atom->nlocal + atom->nghost;
```
Hier auch lieber nmax verwenden.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -0,0 +1,402 @@
+void FixChargeUpdate::pre_force(int) {
+  std::vector<int> mpos = local_to_matrix();
+  int const nall = atom->nlocal + atom->nghost;
+  double **a = read_matrix ? matrix_from_file : array_compute->array;
```
Das verstehe ich noch nicht ganz. Ist das der Grund warum du array in compute_conp_matrix.cpp brauchst?

ludwig-ahrens replied:
Ja, dadurch dass array zu compute.h gehört, kann ich vom fix aus darauf zugreifen.

robeme replied:
Ah OK! Habe ich mir schon gedacht! Kannst du dann nicht gradQ_V und array zusammenfassen? Eigentlich brauchen wir ja keine zwei Matrizen oder? 

array_compute würde ich dann in matrix_compute->array umbennen?

ludwig-ahrens replied:
array ist aber die invertierte gradQ_V. Ich habe auch schon überlegt, die nicht invertierte Matri an den Fix zu übergeben, dann hätte man die Möglichkeit die uninvertierte Matrix auszugeben

robeme replied:
Ich würde das begrüßen. Wäre etwas sauberer. Denn nur der fix braucht ja eigentlich die invertierte Matrix, deshalb solte die eigentlich auch erst dort invertiert werden. Siehst du das ähnlich?

ludwig-ahrens replied:
Ja, ich pass das an

ludwig-ahrens committed:
Cleanup

robeme commented on src/USER-CONP/compute_conp_matrix.cpp:
```diff
@@ -487,14 +485,14 @@ void ComputeConpMatrix::matrix_assignment() {
   MPI_Allgatherv(itaglist_local, igroupnum_local, MPI_LMP_TAGINT, itaglist,
                  igroupnum_list, idispls, MPI_LMP_TAGINT, world);
 
   // sort individual group taglists, first igroup than jgroup
```
Du könntest die Kommentare auch gerne anpassen :)

robeme commented on src/USER-CONP/compute_conp_matrix.cpp:
```diff
@@ -91,11 +91,10 @@ ComputeConpMatrix::ComputeConpMatrix(LAMMPS *lmp, int narg, char **arg)
 
   // TODO recalculate coulomb matrix every recalc_every
```
Hier könnte der Kommentar auch durchaus entfernt werden :)

ludwig-ahrens committed:
Cleanup comments

robeme commented on src/USER-CONP/compute_conp_vector.cpp:
```diff
@@ -217,37 +191,25 @@ void ComputeConpVector::init() {
-  igroupnum = group->count(igroup);
-  jgroupnum = group->count(jgroup);
-  ngroup = igroupnum + jgroupnum;
+  ngroup = group->count(igroup);
```
Test

ludwig-ahrens committed:
Invert capacitance in fix instead of compute

Move invert() method from compute_conp_matrix to fix_charge_update.
This will allow reading and writing the uninverted matrix from the fix.

ludwig-ahrens committed:
Option to read and write capacitance matrix

robeme commented on src/USER-CONP/compute_conp_matrix.cpp:
```diff
@@ -503,7 +457,7 @@ void ComputeConpMatrix::matrix_assignment() {
   // create global matrix indices for local+ghost atoms
   for (bigint ii = 0; ii < ngroup; ii++)
-    for (int i = 0; i < nlocal; i++) // TODO should this be nmax?
+    for (int i = 0; i < nlocal; i++)  // TODO should this be nmax?
```
Warum sollte es nmax sein?

ludwig-ahrens replied:
Weil vielleicht nicht alle Elektrodenatome (also ngroup) in nlocal sind und dann hat mat2tag Lücken

robeme replied:
mat2tag hat definitv Lücken. Deswegen habe ich die Lücken auf -1 gesetzt.

ludwig-ahrens replied:
In pair_contribution() wird das mit tag[j] verglichen und das j kommt aus der Nachbarsliste. D.h. das passende mat2tag könnte da eine Lücke haben.

robeme replied:
Ah, OK, jetzt verstehe ich was du meinst. Ne, sollte alles in Ordnung sein, so lange j in tag[j] kleiner als nlocal ist. An der Stelle wollen wir auch nur die lokalen Atome der Domäne betrachten. Das mit nlocal und nghost hatte ich nur ma' geschrieben, weil es die Matrixzuordnung für die Nachbaratome im SR Teil leichter gemacht hat, wenn j ein Ghost Atom war. War aber blöde für den Rest. Ist jetzt etwas umständlich in SR implentiert, weil der SR Teil aber so schnell geht, sollte das kein Problem sein.

robeme replied:
Im Prinzip könnte man auch ein zusätzliches Atom Property machen (wie zB eine per-Atom Masse; gibt's ein Beispiel in LAMMPS) die dann auch für Ghosts kommuniziert wird. Nur das dann dort halt der Matrixindex oder halt -1 steht.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -16,14 +19,20 @@
-  matrix_from_file = nullptr;
-  read_matrix = false;
+  f_ela = f_cap = f_vec = nullptr;
+  read_elas = read_cap = false;
```
Eigentlich find ich f_mat und f_inv etwas einleuchtender oder was meinst du?

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -2,7 +2,10 @@
+#include <fstream>
+#include <iostream>
 #include <numeric>
+#include <sstream>
```
Sind das alles Standardheader, die so auch auf jedem Rechner verfügbar sein sollten?

ludwig-ahrens replied:
Ich denke schon. Das gehört zur gleichen Input/Output-Bibliothek wie auch iostream und fstream

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -64,17 +79,17 @@ FixChargeUpdate::FixChargeUpdate(LAMMPS *lmp, int narg, char **arg)
+        read_cap = true;
+        input_file_cap = arg[++iarg];
+      } else {
+        error->all(FLERR, "Illegal fix update_charge command");
```
Sollte es hier nicht eher "read elas" bzw. "read coul" sein?

robeme replied:
Oder auch einfach "read mat" bzw. "read inv"...

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -108,7 +123,7 @@ FixChargeUpdate::FixChargeUpdate(LAMMPS *lmp, int narg, char **arg)
   // error checks
-  if (array_compute == nullptr && !read_matrix) {
+  if (array_compute == nullptr && !(read_elas || read_cap)) {
     error->all(FLERR, "Fix charge_update needs one array compute");
```
Ich würde es hier als Fehlermeldung schon matrix nennen. Intern müssen wir ja leider die Bezeichnung array verwenden.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -121,15 +136,16 @@ FixChargeUpdate::FixChargeUpdate(LAMMPS *lmp, int narg, char **arg)
     }
   }
+  if (read_elas && read_cap)
+    error->all(FLERR, "Cannot read matrix from two files");
```
Könnte ja aber durchaus sinnvoll sein, wenn man nicht invertieren möchte oder so?

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -121,15 +136,16 @@ FixChargeUpdate::FixChargeUpdate(LAMMPS *lmp, int narg, char **arg)
+    error->all(FLERR, "Cannot read matrix from two files");
+  if (f_cap && read_elas)
+    error->all(FLERR,
+               "Cannot write capacitance matrix if reading elastance matrix "
```
Ich glaube es heißt coulomb matrix und nicht capacitance matrix

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -173,23 +189,50 @@ void FixChargeUpdate::setup(int) {
 
   // setup pots with target potentials
   std::vector<int> mpos = local_to_matrix();
+  pots = std::vector<double>(ngroup);
```
ich würde phi oder psi besser finden als pots.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -173,23 +189,50 @@ void FixChargeUpdate::setup(int) {
-    read_from_file();
+  std::vector<std::vector<double>> capacitance;
+  if (read_elas) {
+    elastance = read_from_file(input_file_elas);
```
->elas_mat?

ludwig-ahrens replied:
Ich finde elastance klarer. Dass das eine Matrix ist, sieht man ja in der Definition.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -173,23 +189,50 @@ void FixChargeUpdate::setup(int) {
+  if (read_elas) {
+    elastance = read_from_file(input_file_elas);
+  } else if (read_cap) {
+    capacitance = read_from_file(input_file_cap);
```
->capa_mat? Oder einfach inv?

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -322,33 +387,35 @@ int FixChargeUpdate::setmask() {
 void FixChargeUpdate::write_to_file(FILE *file, std::vector<tagint> tags,
                                     std::vector<std::vector<double>> mat) {
   fprintf(file, "# atoms\n");
```
Lass uns die #-Kommentare weglassen.

robeme commented on src/USER-CONP/fix_charge_update.cpp:
```diff
@@ -322,33 +387,35 @@ int FixChargeUpdate::setmask() {
-          tags.push_back(atoi(word));
-          word = strtok(NULL, " \t");
+        while (std::getline(stream, word, ' ')) {
+          if (word == "") continue;
```
Was macht das "if (word == "")" hier genau?

ludwig-ahrens replied:
Wenn mehrere Leerzeichen hintereinander stehen, gibt mir die Schleife leere Strings, die ich so überspringe

ludwig-ahrens committed:
Clearer names in fix charge_update

ludwig-ahrens committed:
Remove TODO
